core/DCache.cpp

```diff
@@ -35,71 +66,246 @@ namespace olympia {
+        ILOG(mem_access_info_ptr << " in read stage");
+        if (incoming_cache_refill_ == mem_access_info_ptr)
+        {
+            reloadCache_(mem_access_info_ptr->getPhyAddr());
+            return;
+        }
+
+        if (mem_access_info_ptr->isCacheHit())
+        {
+            mem_access_info_ptr->setDataReady(true);
+        }
+        else
+        {
+            if (!l2cache_busy_)
+            {
+                out_l2cache_req_.send(mem_access_info_ptr);
+                l2cache_busy_ = true;
+            }
+            else
+            {
+                uev_mshr_request_.schedule(sparta::Clock::Cycle(1));
+            }
+        }
+        out_lsu_lookup_ack_.send(mem_access_info_ptr);
+    }
+
+    void DCache::mshrRequest_()
+    {
+        if (!l2cache_busy_)
+        {
+            auto iter = mshr_file_.begin();
+            while (iter != mshr_file_.end())
+            {
+
+                if (iter.isValid())
+                {
+                    const auto & mshr_entry = *iter;
+                    auto mem_info = mshr_entry->getMemRequest();
+                    if (mshr_entry->isValid() && !mshr_entry->getDataArrived() && mem_info)
+                    {
+                        ILOG("Sending mshr request when not busy " << mem_info);
+                        out_l2cache_req_.send(mem_info);
+                        l2cache_busy_ = true;
+                        break;
+                    }
+                }
+                ++iter;
+            }
+        }
+    }
+
+    void DCache::handleDeallocate_()
+    {
+        const auto stage_id = static_cast<uint32_t>(PipelineStage::DEALLOCATE);
+        const MemoryAccessInfoPtr & mem_access_info_ptr = cache_pipeline_[stage_id];
+        ILOG(mem_access_info_ptr << " in deallocate stage");
+        if (incoming_cache_refill_ == mem_access_info_ptr)
+        {
+            ILOG("Refill complete " << incoming_cache_refill_);
+            incoming_cache_refill_.reset();
+            const auto & mshr_it = mem_access_info_ptr->getMSHRInfoIterator();
+            if (mshr_it.isValid())
+            {
+                MemoryAccessInfoPtr dependant_load_inst = (*mshr_it)->getMemRequest();
+                out_lsu_lookup_ack_.send(dependant_load_inst);
+
+                ILOG("Removing mshr entry for " << mem_access_info_ptr);
+                mshr_file_.erase(mem_access_info_ptr->getMSHRInfoIterator());
+            }
+            return;
+        }
+        ILOG("Deallocating pipeline for " << mem_access_info_ptr);
+    }
+
+    void DCache::getMemReqFromLSU_(const MemoryAccessInfoPtr & memory_access_info_ptr)
+    {
+
+        ILOG("Got memory access request from LSU " << memory_access_info_ptr);
+        if (!cache_refill_selected_)
+        {
+            ILOG("Arbitration from refill " << memory_access_info_ptr);
+            memory_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::MISS);
```

Comment:
It's not really a miss, but a loss of arbitration.  Does the LSU know to try again next cycle?

Reply:
I have replaced it with RELOAD as i think its also used in the l2 cache for a similar situation